### ISSUE TITLE
Getting a dispersive material with epsilon_input_file or adding new geometry object

### ISSUE THREAD

**Issue by MolinAlexei** [NanoComp/meep#1504]
Hi,

I am using MEEP for a 3D two lensed optical system operating in the mm wavelengths. In order to create the lenses I use the epsilon_input_file in Sim, to which I give a "dielectric map" of my lenses, so I can control precisely their non trivial shape. 
However I need to model an absorbing tube (with a high loss tangent) around the lenses (like a telescope), which requires a material with a complex permittivity. The espilon_input_file can only be real valued positive permittivities, meaning I can't use this for that purpose. I can however create the tube using objects from geometry (with a block of material and a cylinder of mp.air) with the desired properties. The problem is, the epsilon_input_file is only used where there are no geometry objects defined, meaning that if I do use geometry objects to create the tube, I cannot put the lenses inside the said tube, eventhough it is really hollow and there is no reason why that shouldn't be possible.

My question is : is there a way to force the Simulation to read the map provided to it where I know there is only air/vacuum ? If not, could it be possible to implement a geometry object as a hollow tube for which MEEP could still accept reading espilon_input_file in the hollow part ?

**Comment by stevengj** [comment]
> The problem is, the epsilon_input_file is only used where there are no geometry objects defined

I don't think this was true?  geometric objects should appear "on top of" the epsilon_input_file, no?

There is a more general "MaterialGrid" type that was used to implement adjoint optimization, which supports dispersive materials.   It needs to be better documented, though we are still discussing the API (#1298).

**Comment by MolinAlexei** [comment]
Thank you for answering,
What I tested agrees with this quote from the documentation (in Simulation epsilon_input_file) : "This dielectric function is then used in place of the ε property of default_material (i.e. where there are no geometry objects)."  Geometry objects simply replace the permittivity given by the map where they are defined.

I didn't know about the MaterialGrid type. From what I've seen in #1298 it's used in the geometry objects which suggests that the same problem would occur : the simulation would still overwrite the geometry objects on the permittivity map because they technically overlap in their definition. I am however perhaps wrong and will look further into it.

**Comment by stevengj** [comment]
Oh, I see what you are saying — Meep doesn't have a "hollow tube" geometric object, and if you do two nested cylinders then the inside of the tube is still "on top of" the lens shape that you want to create with the epsilon input file.     In principle, a MaterialGrid may allow you to get around this because you can apply the grid to an object that you put "on top" of the tube objects.

Actually, if I remember correctly, there is an undocumented feature where you can simply set the "material" of an object to be the name of a file, and then it will interpolate within that file for your material.   That is, your geometry will consist of a cylinder object for your lossy tube material, then another "inner" cylinder for your lens — but the `material` of the inner cylinder will be set to the name of your input file (i.e. a *string*).

**Comment by MolinAlexei** [comment]
Thanks you again for answering ! 

I tried to change the material in a geometry object by a string, where the string is the name of the permittivity map for my lenses. If the geometry object is a cylinder, the process seems to freeze just after displyaing the different objects with their properties, which is where subpixel averaging would occur. If the object is a block (which is not a telescope tube but I tried just to see if it worked), the technique seems to work, however for some reason the map is only read from about 2/3 of its axis 0 (the map is 3D and what I call the axis 0 is just the first coordinate, like an array). The process works but as if the map was empty for the first 2/3s of the axis 0 and only reads it after that. 

I haven't tried the MaterialGrid option yet, I will keep you informed.

**Comment by stevengj** [comment]
Thanks for trying this; it seems like it would be worth getting this to work in general and adding a test case, but it might require some debugging since I don't recall whether we've actually used it in this way — currently it is just a quirk of how `epsilon_input_file` was implemented in meepgeom.cpp as a special kind of `default_material`.

**Comment by MolinAlexei** [comment]
I tried the MaterialGrid; it doesn't work when applied on a cylinder of course, however it does work on a block. This means that I have to make a map with the lenses and some tube around them. 
 The only problem is that it doesn't seem to account for complex values (that I need to get absorption on the tube). What I understood is that there are two materials specified and a map with elements ranging from 0 to 1. Then MaterialGrid interpolates a material between the two specified materials depending on the value at the given point. Please correct me if I am wrong. I could make MaterialGrid interpolate between mp.air and a material with a complex permittivity, but it wouldn't serve the purpose I need : the lenses need to be purely dielectric, and the tube absorbing. This means that there are 3 materials : air/vacuum, dielectric, and complex permittivity for absorbtion; meaning whatever is done by MaterialGrid cannot work.

Let me know if there is anything that can be done. I haven't been looking into MEEP's code itself as I am probably not comfortable enough with this kind of code to do anything useful for my problem.
